src/py/flwr/common/backoff.py

```diff
@@ -0,0 +1,278 @@
+        Factor by which the delay is multiplied after each retry.
+    max_delay: Optional[float] (default: None)
+        The maximum delay duration between two consecutive retries.
+    """
+    delay = base_delay if max_delay is None else min(base_delay, max_delay)
+    while True:
+        yield delay
+        delay *= multiplier
+        if max_delay is not None:
+            delay = min(delay, max_delay)
+
+
+def constant(
+    interval: Union[float, Iterable[float]] = 1,
+) -> Generator[float, None, None]:
+    """Wait time generator for specified intervals.
+
+    Parameters
+    ----------
+    interval: Union[float, Iterable[float]] (default: 1)
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def random_jitter(value: float) -> float:
+    """Jitter the value by a random number of milliseconds.
+
+    This adds up to 1 second of additional time to the original value.
+    Prior to backoff version 1.2, this was the default jitter behavior.
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return value + random.random()
+
+
+def full_jitter(value: float) -> float:
+    """Jitter the value across the full range (0 to value).
+
+    This corresponds to the "Full Jitter" algorithm specified in the
+    AWS blog post on the performance of various jitter algorithms.
+    See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return random.uniform(0, value)
+
+
+@dataclass
+class Details:
+    """Details for event handlers in SafeInvoker."""
+
+    func: Callable[[Any], Any]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    tries: int
+    elapsed: float
+    exception: Optional[Exception] = None
+    wait: Optional[float] = None
+
+
+# pylint: disable-next=too-many-instance-attributes
+class SafeInvoker:
+    """Wrapper class for backoff and retry triggered by exceptions.
+
+    Parameters
+    ----------
+    wait: Generator[float, None, None]
+        A generator yielding successive wait times in seconds. If the generator
+        is finite, the giveup event will be triggered when the generator raises
+        `StopIteration`.
+    exception:  Union[Type[Exception], Iterable[Type[Exception]]]
+        An exception type (or iterable of types) that triggers backoff.
+    max_tries: Optional[int] (default: None)
+        The maximum number of attempts to make before giving up. Once exhausted,
+        the exception will be allowed to escape. If set to None, there is no limit
+        to the number of tries.
+    max_time: Optional[float] (default: None)
+        The maximum total amount of time to try before giving up. Once this time
+        has expired, the exception will be allowed to escape.
```

Comment:
Let's document the meaning of `None` here as well.

Reply:
Done